dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSchema.h

```diff
@@ -0,0 +1,140 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <Common/nocopyable.h>
+#include <Core/Block.h>
+#include <Interpreters/Context.h>
+#include <Storages/BackgroundProcessingPool.h>
+#include <Storages/DeltaMerge/DeltaMergeDefines.h>
+#include <common/types.h>
+#include <openssl/base.h>
+#include <openssl/sha.h>
+
+#include <boost/container_hash/hash_fwd.hpp>
+
+namespace std
+{
+using Digest = UInt256;
+template <>
+struct hash<Digest>
+{
+    size_t operator()(const Digest & digest) const
+    {
+        size_t seed = 0;
+        boost::hash_combine(seed, boost::hash_value(digest.a));
+        boost::hash_combine(seed, boost::hash_value(digest.b));
+        boost::hash_combine(seed, boost::hash_value(digest.c));
+        boost::hash_combine(seed, boost::hash_value(digest.d));
+        return seed;
+    }
+};
+} // namespace std
+
+namespace DB
+{
+namespace DM
+{
+using Digest = UInt256;
+class ColumnFileSchema
+{
+private:
+    Block schema;
+
+    using ColIdToOffset = std::unordered_map<ColId, size_t>;
+    ColIdToOffset colid_to_offset;
+
+public:
+    explicit ColumnFileSchema(const Block & schema_)
+        : schema(schema_)
```

Comment:
How about always call `schema_.cloneEmpty()`? I guess it will not be slower than copying a `schema_` while it provides nice features, for example, you can just pass any existing block to it.